[김성태] 작품 메인페이지 엔드포인트 (feature/works_main)

kimst5418 opened the pull request:
## 수정 사항 간략한 한줄 요약
- 수정요청 사항 반영

## 수정 사항들 자세한 내용
(여기에 수정 사항에 대한 자세한 내용을 작성해 주세요.)
- 리뷰내용 반영
- WorksListView 에서 토큰 데코레이터 추가 (데코레이터가 다른 branch에 있긴한데 합쳐서 테스트 했을땐 정상작동 했습니다.)
- WorksListView 에 카테고리별 필터와 팔로우한 크리에이터의 작품 필터 추가

## 기타 질문 및 특이 사항
(궁금한 사항들, 하면서 느낀 점들 공유해주실 것이 있으면 자유롭게 작성해 주세요.)

## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.

kimst5418 committed:
Add - 작품 main 화면 및 카테고리 인기크리에이터 엔드포인트 추가
- 작품 main 화면피드 부분 엔드포인트 추가
- 작품 카테고리 페이지 태그 엔드포인트 추가
- 작품 카테고리 페이지 인기크리에이터 엔드포인트 추가

kimst5418 committed:
Modify : 인기크리에이터 정렬 조건 수정
- 인기크리에이터 정렬을 해당 크리에이터의 모든 게시글의 좋아요 수에서 해당 카테고리 게시글의 좋아요 수 합산으로 변경

kimst5418 committed:
Add : 작품 메인화면 카테고리 리스트 엔드포인트 구현
- 작품 메인화면 카테고리 리스트 카테고리명과 해당 카테고리 포스팅 수를 주는 view 구현

kimst5418 committed:
modify : CategoryTagView 에러처리 수정

kimst5418 committed:
workslist 데뷰 필터 추가

lordmyshepherd commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+    ReplyLike
+)
+
+class WorksListView(View) :
```
성태님! WorksListView의 역할이 "최신", ""주목받는", "발견", "데뷰" 조건에 맞는 work 목록을 반환하는게 맞나요?

kimst5418 replied:
네 맞습니다.

lordmyshepherd commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+                "Comments"      : work.comment_set.count(),
+                "Views"         : work.views,
+                "singup_time"   : work.user.created_at
+            } for work in work_all ]
```
work_all = Work.objects.all()이면 이후에 work_all을 for문 돌릴 경우, queryset이 평가된다고 합니다.
이 때, `work에 table에 column들은 처음에 가져오지만, work.user.user_name으로 다른 테이블에 접근 할 경우 for문이 실행될 때마다 데이터베이스에 조회하게 됩니다.`
이 경우에, select_related() 또는 prefetch_related() method를 사용하면 table을 Ieft outer join문을 사용해서 한번에 모든 데이터를 가져오기 때문에 처음 한번만 데이터베이스를 hit하고 나머지는 값들은 메모리에 저장해 놓고 불러올 수 있습니다!
한번, 사용해보시는게 어떨까요?

kimst5418 replied:
반영하였습니다.

kimst5418 replied:
반영하였습니다.

soheon-lee commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+
+class WorksListView(View) :
+
+    def get(self, request, sort, page) :
```
들어오는 sort 이름과 order_by할 조건은 딕셔너리의 형태로 미리 선언해두시면 키:밸류로 접근이 가능하기 때문에 `if와 `elif`를 많이 줄일 수 있을 것 같습니다 :)

kimst5418 replied:
반영하였습니다

lordmyshepherd commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+        categorylist = [ {
+            "categoryName"  : category.name,
+            "categoryCount" : category.work_set.count()
+        } for category in Category.objects.all() ]
```
여기도 마찬가지 입니다.

kimst5418 replied:
반영하였습니다.

lordmyshepherd commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+
+class CategoryTagView(View) : 
+
+    def get(self, request, category) :
```
성태님 여기서 category 이름보다는 category_id로 통신하는 것이 좋습니다!

kimst5418 replied:
category_id 로 수정하였습니다.

lordmyshepherd commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+
+    def get(self, request, category) :
+        try :
+            category_id = Category.objects.get( name = category ).id
```
위와 같은 얘기로 category_id로 통신하게되면 get() filter를 각 각 보낼 필요가 없어 보입니다~
한번 수정해보세요~~

kimst5418 replied:
category_id 로 수정하였습니다.

lordmyshepherd commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+
+    def get(self, request, sort, page) :
+        if sort == "최신" :
+            work_all = Work.objects.all().order_by('-created_at')
```
22~29 부분을 query string을 활용해서 구현해보세요.
또한 다중 filter를 구현하는 경우 Q객체를 사용해서 구현해보시면 좋을 것 같습니다

kimst5418 replied:
query string 반영하였습니다.

soheon-lee commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+            workslist=sorted(workslist, reverse=True, key=lambda x: x["Likes"])[12*(page-1):12*page]
+            return JsonResponse({'data': workslist }, status=200)
+        elif sort == "데뷰" :
+            workslist=sorted(workslist, reverse=True, key=lambda x: x["singup_time"])[12*(page-1):12*page]
```
lambda good !

soheon-lee commented on work/views.py:
```diff
@@ -0,0 +1,91 @@
+    def get(self, request, sort, page) :
+        if sort == "최신" :
+            work_all = Work.objects.all().order_by('-created_at')
+            work_all = work_all[12*(page-1):12*page]
```
한 페이지에 나올 데이터 수가 `12`개인가요?
이렇게 일일이 적어두시는 것은 하드코딩입니다. 함수 상단에서 Constant로 지정해주시거나, 아니면 서비스의 확장성을 위해서 프론트엔드에서 숫자를 받아오시는 것도 한 방법입니다. 서비스의 기획이 변경되어 한 페이지에 15장씩 보여주게 되었을 때, 프론트엔드에서 자유롭게 그 수를 변경할 수 있도록 자유도를 주는 것이 조금 더 확정성있기 때문입니다.

kimst5418 replied:
offset, limit 방식으로 수정하였습니다.

kimst5418 committed:
Modify : 리뷰 반영

kimst5418 committed:
Modify : conflict 수정

kimst5418 committed:
Modify : 리뷰반영

kimst5418 committed:
Modify : category view image_url 추가

kimst5418 committed:
Modify : 오타수정

soheon-lee commented on work/views.py:
```diff
@@ -1 +1,100 @@
+
+        if sort in sort_order :
+            works = Work.objects.all().order_by(sort_order[sort]).select_related("user").prefetch_related("workimage_set", "likeit_set", "comment_set")[offset:(offset+limit)]
+        elif sort == "발견" or sort == "데뷰" :
+            works = Work.objects.all().select_related("user").prefetch_related("workimage_set", "likeit_set", "comment_set")
```
"발견" 일 때나 "데뷰" 일 때는 order_by 조건을 ('id')로 하시면 좋을 것 같습니다. 그러면 코드가 중복되지 않을 수도 있고, 결과도 똑같이 나오게 됩니다 :)

kimst5418 replied:
id 조건을 어떻게 주면 좋을지 아이디어가 없어서 dictionary 를 수정하여 코드를 수정하였습니다.
코드는 줄였는데 카테고리별 필터와 '피드'라고 내가 팔로우한 creator의 작품을 보여주는 분류도 추가하다 보니
코드가 다시 길어졌습니다... 확인부탁드립니다.

soheon-lee commented on work/views.py:
```diff
@@ -1 +1,100 @@
+            return JsonResponse({'MESSAGE':'Invalid sorted name'}, status=400)
+
+        workslist = [ {
+                "id"            : work.id ,
+                "AuthorName"    : work.user.user_name ,
+                "AuthorProfile" : work.user.profile_image_url ,
+                "PostName"      : work.title ,
+                "Img"           : work.workimage_set.first().image_url,
+                "Likes"         : work.likeit_set.count(),
+                "Comments"      : work.comment_set.count(),
+                "Views"         : work.views,
+                "singup_time"   : work.user.created_at
```
```suggestion
                "id"            : work.id,
                "AuthorName"    : work.user.user_name,
                "AuthorProfile" : work.user.profile_image_url,
                "PostName"      : work.title,
                "Img"           : work.workimage_set.first().image_url,
                "Likes"         : work.likeit_set.count(),
                "Comments"      : work.comment_set.count(),
                "Views"         : work.views,
                "singup_time"   : work.user.created_at
```

kimst5418 replied:
반영하였습니다.

soheon-lee commented on work/views.py:
```diff
@@ -1 +1,100 @@
+                "Views"         : work.views,
+                "singup_time"   : work.user.created_at
+            } for work in works ]
+        if sort == "발견" :
+            workslist=sorted(workslist, reverse=True, key=lambda x: x["Likes"])[offset:(offset+limit)]
+            return JsonResponse({'data': workslist }, status=200)
+        elif sort == "데뷰" :
+            workslist=sorted(workslist, reverse=True, key=lambda x: x["singup_time"])[offset:(offset+limit)]
+            return JsonResponse({'data': workslist }, status=200)
```
성태님 여기도 같은 코드가 반복되는데, 조금 더 연구하셔서 여기 한 줄로 붙이시는게 좋을 것 같습니다 👍

kimst5418 replied:
한줄로 수정하였습니다.

soheon-lee commented on work/views.py:
```diff
@@ -1 +1,100 @@
+                "follower"      : user.user_to_follow.count(),
+                "like"          : sum([ works.likeit_set.count() for works in user.work_set.all() ]),
+                "illust"        : user.work_set.count(),
+                "imgPreviewSrc" : [works.workimage_set.all()[0].image_url for works in user.work_set.all()][:3],
```
Django에는 쿼리셋 요소 가운데 첫번째 요소를 가져올 수 있는 메소드가 있습니다 :) 한 번 활용해보세요. all()을 사용하는 것보다 시간이 조금이나마 덜 소요됩니다.

kimst5418 replied:
first() 로 반영하였습니다.

kimst5418 committed:
Modify : workslistview 리뷰반영
- workslistview 리뷰반영
- workslistview에 팔로잉 작품별 필터 추가
- workslistview에 카테고리별 필터기능 추가

soheon-lee commented:
LGTM
